File: R/multiple_linear_regression.R
#Regresion Lineal multiple

#Cargar Dataset
dataset = read.csv('50_Startups.csv')

#Variables Dummy
dataset$State =  factor(dataset$State,
                        levels= c("New York", "California", "Florida"),
                        labels = c(1,2,3))

library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8) 
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)


#AJUSTAR EL MODELO DE REGRESION LINEAL MULTIPLE CON EL MODELO, 
#El punto hace referencia a que esta usando todas las variables independientes en lugar de escribirlas
#todas se pone un .
regression= lm(formula = Profit ~ .,
               data = training_set)


#Predecir los resultados con el conjunto testing
y_pred = predict(regression, newdata = testing_set)

#Eliminacion hacia atras
SL=0.05
regression = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
                data = dataset)
summary(regression)

regression = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend ,
                data = dataset)
summary(regression)

regression = lm(formula = Profit ~ R.D.Spend + Marketing.Spend ,
                data = dataset)
summary(regression)

regression = lm(formula = Profit ~ R.D.Spend,
                data = dataset)
summary(regression)

File: Python/multiple_linear_regression.py

#Regresion Lineal multiple
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Cargar el dataser
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:,4].values

#Datos Categoricos
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X= LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:,3])
onehotencoder = OneHotEncoder(categorical_features = [3])
X= onehotencoder.fit_transform(X).toarray()

#Ajuste de Variables Dummy
X= X[:, 1:]

# Split de Train y Test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

#Ajustar el Modelo de Regresion lineal multiple al conjunto de entrenamiento 
from sklearn.linear_model import LinearRegression
regression = LinearRegression()
regression.fit(X_train, y_train)

#Prediccion de Resultados
y_pred = regression.predict(X_test)

#Eliminción hacia atras
import statsmodels.api as sm
X = np.append(arr= np.ones((50,1)).astype(int), values = X, axis = 1)

SL = 0.05

X_opt = X[:, [0,1,2,3,4,5]]
regression_OLS = sm.OLS(endog =y, exog = X_opt).fit()
regression_OLS.summary()

X_opt = X[:, [0,1,3,4,5]]
regression_OLS = sm.OLS(endog =y, exog = X_opt).fit()
regression_OLS.summary()

X_opt = X[:, [0,3]]
regression_OLS = sm.OLS(endog =y, exog = X_opt).fit()
regression_OLS.summary()

#Podemos concluir que la variable de la columna 3 es la mejor para hacer 
#la prediccion y nos lleva a que podria ser una regresion lineal simple
#Sin embargo existen ma metodos los cuales podemos usar elegir la variable 
#Se puede automatizar esto de ir repitiendo lineas de codigo con un ciclo
# o una funcion que lo haga por nosotros






